File: Assignment5/BabyNames.go
/**
* An implementation of the "Baby Names" program in Go where, when the user
* gives a name or set of names, checks a "database" (a set of files with
* information on names for many years) and returns information about
* that name if it finds it as well as for names that contain that given name
* as a prefix.
* author @ayang
* Created: Dec 5, 2021
 */

package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
)

type BabyName struct {
	Name   string
	Amount int
	Years  int
}

// ByName implements sort.Interface based on the Name field
type ByName []BabyName

func (name ByName) Len() int           { return len(name) }
func (name ByName) Less(i, j int) bool { return name[i].Name < name[j].Name }
func (name ByName) Swap(i, j int)      { name[i], name[j] = name[j], name[i] }

/*
	Modified version of the readHurricane code provided by Prof. Towell,
	changed so that it returns a string slice that contains each row of the
	excel sheet as an element, ideally

	@param url (the url to get the information from), maxLines (the amount of lines you want read)

	@return nameLst (a slice containing all information from the specified amount of lines), nil iff
	there was an error in reading the url
*/
func readName(url string, maxLines int) []string {
	resp, err := http.Get(url)
	// if there is a problem with the url, error message
	if err != nil {
		fmt.Println("Http problem", err)
		return nil
	}

	// read all the text in the url and then close
	body, err := io.ReadAll(resp.Body)
	resp.Body.Close()
	// if there is a problem with reading the text in the url, error message
	if err != nil {
		fmt.Println("Http problem2", err)
		return nil
	}

	// convert the text read into a string
	str := string(body)
	r := csv.NewReader(strings.NewReader(str))
	r.FieldsPerRecord = -1

	var nameLst []string

	// store the value for each line
	for i := 0; i < maxLines || maxLines <= 0; i++ {
		// for every line read it
		record, err := r.Read()

		// if there is no more input to read, stop
		if err == io.EOF {
			break
		}
		// if there is an error, print out the error and stop
		if err != nil {
			fmt.Fprintf(os.Stderr, "%v\n", err)
			break
		}
		name := strings.Join(record, ",")
		nameLst = append(nameLst, name)
		// return string???
	}
	return nameLst
}

/*
	Store names from a string (from a slice) into two separate slices based on whether the name is
	masculine or feminine usage

	@param: name (section from slice to sort), boyLst and girlLst (slices for masc/femme usage respectively)
	@return: updated boyLst and girlLst (bc Go is something something referenced based so I don't have fun :/)
*/
func nameStorer(name string, boyLst []BabyName, girlLst []BabyName) ([]BabyName, []BabyName) {
	nameLst := strings.Split(name, ",") // make the row into a slice of strings split at commas

	// get the amounts for the names. BEFORE anything else because I will use this A Lot
	boyAmount, err := strconv.Atoi(strings.TrimSpace(nameLst[2]))
	if err != nil {
		fmt.Fprintf(os.Stdout, "CONV PROB [[%v]]  %v\n", nameLst[2], err)
	}
	girlAmount, err := strconv.Atoi(strings.TrimSpace(nameLst[4]))
	if err != nil {
		fmt.Fprintf(os.Stdout, "CONV PROB [[%v]]  %v\n", nameLst[4], err)
	}

	// if the list is empty, just add the name. since boy & girl list are  the same length uh some questionable things are being done
	if len(boyLst) == 0 && len(girlLst) == 0 {
		// add the item to the list
		boyLst = append(boyLst, BabyName{nameLst[1], boyAmount, 1})
		girlLst = append(girlLst, BabyName{nameLst[3], girlAmount, 1})
		return boyLst, girlLst
	}

	// check to see if an instance of that name exists already
	boyIndx := -1
	girlIndx := -1
	for i := 0; i < len(boyLst); i++ {
		if strings.EqualFold(nameLst[1], boyLst[i].Name) {
			boyIndx = i // if there is an instance of the name in the list already, just update it.
			break
		}
	}
	for i := 0; i < len(girlLst); i++ {
		if strings.EqualFold(nameLst[3], girlLst[i].Name) {
			girlIndx = i // if there is an instance of the name in the list already, just update it
			break
		}
	}

	// all combinations of cases of already exists/doesn't already exist
	if boyIndx != -1 && girlIndx != -1 {
		// in the case where both already exist, just update them both
		boyLst[boyIndx].Amount = boyLst[boyIndx].Amount + boyAmount
		boyLst[boyIndx].Years = boyLst[boyIndx].Years + 1

		girlLst[girlIndx].Amount = girlLst[girlIndx].Amount + girlAmount
		girlLst[girlIndx].Years = girlLst[girlIndx].Years + 1

		sort.Sort(ByName(boyLst))
		sort.Sort(ByName(girlLst))

		return boyLst, girlLst
	} else if boyIndx != -1 && girlIndx == -1 {
		// in the case where boy already exists and girl doesn't update boy and add/sort girl
		boyLst[boyIndx].Amount = boyLst[boyIndx].Amount + boyAmount
		boyLst[boyIndx].Years = boyLst[boyIndx].Years + 1

		girlLst = append(girlLst, BabyName{nameLst[3], girlAmount, 1})

		sort.Sort(ByName(boyLst))
		sort.Sort(ByName(girlLst))

		return boyLst, girlLst
	} else if boyIndx == -1 && girlIndx != -1 {
		// in teh case where girl already exists and boy doesn't update girl and add/sort boy
		boyLst = append(boyLst, BabyName{nameLst[1], boyAmount, 1})

		girlLst[girlIndx].Amount = girlLst[girlIndx].Amount + girlAmount
		girlLst[girlIndx].Years = girlLst[girlIndx].Years + 1

		sort.Sort(ByName(boyLst))
		sort.Sort(ByName(girlLst))

		return boyLst, girlLst
	} else {
		// in the case where both girl and boy doesn't exist, add and sort for both
		boyLst = append(boyLst, BabyName{nameLst[1], boyAmount, 1})
		girlLst = append(girlLst, BabyName{nameLst[3], girlAmount, 1})

		sort.Sort(ByName(boyLst))
		sort.Sort(ByName(girlLst))

		return boyLst, girlLst
	}
}

/*
	Given a name and slices to search from, check to see if that given name (and any name that contains the given name
	as a prefix) exist in the list. If so, print out the name along with some statistics.

	@param: name (name you are looking for instances of), boyLst and girLst (lists you are searching for instances of "name")
*/
func nameSearch(name string, boyLst []BabyName, girlLst []BabyName) {
	var boyIndxs []int  // slice of indexes for the name to print information about for masc names
	var girlIndxs []int // slice of indexes for the name to print information about for femme names

	// check to see if the name itself is in the slice
	checkName := func(name string, lst []BabyName, lstModify []int) []int {
		for i := 0; i < len(lst); i++ {
			if strings.EqualFold(name, lst[i].Name) {
				lstModify = append(lstModify, i)
			}
		}
		return lstModify
	}

	girlIndxs = checkName(name, girlLst, girlIndxs)
	boyIndxs = checkName(name, boyLst, boyIndxs)

	// check for prefixes
	checkPrefix := func(name string, lst []BabyName, lstModify []int) []int {
		for i := 0; i < len(lst); i++ {
			// dodging the issue of "name you are comparing to is too small to subslice"
			if len(lst[i].Name) >= len(name) {
				// take a substring from the beginning of the size of the name you are comparing to
				prefName := lst[i].Name[0:len(name)]
				if strings.EqualFold(name, prefName) {
					// enter: a convoluted way to avoid duplicates where I don't just write a contains method
					hasName := false // assume it isn't in the ret list
					for j := 0; j < len(lstModify); j++ {
						if i == lstModify[j] {
							hasName = true // if it is in the ret list, change the traacker
						}
					}
					if !hasName {
						// if it is an prefix for the name from the slice, add it to the list
						// fmt.Println("in prefix checker")
						// fmt.Println(i)
						lstModify = append(lstModify, i)
					}
				}
			}
		}
		return lstModify
	}

	girlIndxs = checkPrefix(name, girlLst, girlIndxs)
	boyIndxs = checkPrefix(name, boyLst, boyIndxs)

	// get the total to calculate percentages with
	boyTotal := 0
	girlTotal := 0

	for i := 0; i < len(boyLst); i++ {
		boyTotal = boyTotal + boyLst[i].Amount
	}
	for i := 0; i < len(girlLst); i++ {
		girlTotal = girlTotal + girlLst[i].Amount
	}

	// return results
	fmt.Println(name) // print out the search results for the name

	// if the slice for the respective gender ISN'T empty, print things out. otherwise, print nothing out
	// for masc names:
	if len(boyIndxs) != 0 {
		// fmt.Println("*Snake voice* im in")
		for i := 0; i < len(boyIndxs); i++ {
			nameToPrint := boyLst[boyIndxs[i]] // so i don't have to write "boyLst[boyIndxs[i]]" a lot because it's cumberson and ugly :(

			// find percentage and rank here
			percent := (float64(nameToPrint.Amount) / float64(boyTotal)) * 100
			rank := boyIndxs[i] + 1

			// printing out the information
			fmt.Printf("%s: %s \t %s: %s \t %s: %d\n", "Name", nameToPrint.Name, "Gender", "Boy", "# of Years", nameToPrint.Years)
			fmt.Printf("%s: %d\n", "Alphabetical Position", rank)
			fmt.Printf("%s: %d\n", "Occurences", nameToPrint.Amount)
			fmt.Printf("%s %s %s: %f\n\n", "Percent of All", "Boys", "Names", percent)
		}
	}
	// for femme names:
	if len(girlIndxs) != 0 {
		for i := 0; i < len(girlIndxs); i++ {
			nameToPrint := girlLst[girlIndxs[i]] // so i don't have to write "boyLst[boyIndxs[i]]" a lot because it's cumberson and ugly :(

			// find percentage and rank here
			percent := (float64(nameToPrint.Amount) / float64(girlTotal)) * 100
			rank := girlIndxs[i] + 1

			// printing out the information
			fmt.Printf("%s: %s \t %s: %s \t %s: %d\n", "Name", nameToPrint.Name, "Gender", "Girl", "# of Years", nameToPrint.Years)
			fmt.Printf("%s: %d\n", "Alphabetical Position", rank)
			fmt.Printf("%s: %d\n", "Occurences", nameToPrint.Amount)
			fmt.Printf("%s %s %s: %f\n\n", "Percent of All", "Girl", "Names", percent)
		}
	}
}

func main() {
	var GirlNames []BabyName // list holding information on girl names
	var BoyNames []BabyName  // list holding information on boy names
	var NameQuery []string   // list of names to return results for

	// NameQuery := [4]string{"Aaron", "Devon", "marlen", "syd"}

	// if the user doesn't give any names to test for, yell at them to give name and then end early.
	if len(os.Args[1:]) < 1 {
		fmt.Println("Please enter name(s) to search for!")
		os.Exit(0)
	}
	// get all user inputs
	for i := 1; i < len(os.Args[1:])+1; i++ {
		NameQuery = append(NameQuery, os.Args[i])
	}

	// lsit of names. man, wish there was a way to collapse this!!
	filesLst := [10]string{"https://cs.brynmawr.edu/cs245/Data/names2000.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2001.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2002.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2003.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2004.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2005.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2006.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2007.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2008.csv",
		"https://cs.brynmawr.edu/cs245/Data/names2009.csv"}

	//list that holds the string for each name
	nameLst := make([][]string, 10)

	// read all the files
	for i := 0; i < len(filesLst); i++ {
		var url string
		url = filesLst[i]
		nameLst[i] = readName(url, 1000)
	}

	// put all the information from the names to their respective lists
	for i := 0; i < len(nameLst); i++ {
		for j := 0; j < len(nameLst[i]); j++ {
			// fmt.Println(nameLst[i][j])
			BoyNames, GirlNames = nameStorer(nameLst[i][j], BoyNames, GirlNames)
			// fmt.Println(BoyNames)
		}
	}

	// run the search for if that name (or if a name containing that name as a prefix) exists in the slices
	for i := 0; i < len(NameQuery); i++ {
		nameSearch(NameQuery[i], BoyNames, GirlNames)
	}
}

File: Assignment4/BabyNames.kt
/**
    A program that, when the user gives a name or set of names, checks a "database" 
    (a set of files with information on names for many years) and returns information about
    that name if it finds it. In addition, returns results for names that contain that given name
    as a prefix.
    
    @author ayang
    Created: Nov 15, 2021
 */

import java.io.*
import java.net.*

/**
    Defines a class NameHolder that contains all the information for a given name
    @param: name (the name), uses (how many babies have that name)
 */
class NameHolder (val name: String, val amount: String) {
    var years = 1 
    var totalAmount: Int = amount.toInt() // making it easier to just parse now bc everything in the excel is string. fun!

    // add a year to the list of years
    fun addYear() {
        years = years + 1
    }

    // adds to the total
    fun addAmount(numToAdd: String) {
        totalAmount = totalAmount + numToAdd.toInt()
    }
}

/**
    Stores the names given in the list for each year into two lists: "GirlNames" and "BoyNames"
    @param: name (row of names to store info about), boyLst and girlLst (lists to store the name information about)
*/
fun nameStorer(name: String, boyLst: MutableList<NameHolder>, girlLst: MutableList<NameHolder>) {
    // turn the name into like. something that I can call indexes on.
    val nameLst: List<String> = name.split(",").map { it.trim() }

    // if the list is empty, just add it. checking for both because. security. could honestly probably just "&&" but. "security"
    if(!boyLst.isNotEmpty()){
        boyLst.add(NameHolder(nameLst[1], nameLst[2]))
    }
    if(!girlLst.isNotEmpty()){
        girlLst.add(NameHolder(nameLst[3], nameLst[4]))
    }

    // do for the boy list
    if(containsName(nameLst[1], boyLst) > -1){
        // if the name is in the list, just update the metrics
        val idx = containsName(nameLst[1], boyLst)
        boyLst[idx].addYear()
        boyLst[idx].addAmount(nameLst[2])
    } 
    else {
        // otherwise, create a new item and add it to the list
        addAlphabetically(NameHolder(nameLst[1], nameLst[2]), boyLst)
    }
    // do for the girl list
    if(containsName(nameLst[3], girlLst) > -1){
        // if the name is in the list, just update the metrics
        val idx = containsName(nameLst[3], girlLst)
        girlLst[idx].addYear()
        girlLst[idx].addAmount(nameLst[4])
    } 
    else {
        // otherwise, create a new item and add it to the list
        addAlphabetically(NameHolder(nameLst[3], nameLst[4]), girlLst)
    }
}

/**
    A function that searches for a given name, and if it or ones that contains it as a prefix
*/
fun nameSearch(name: String, lst: MutableList<NameHolder>, gender: String) {
    // search for the name & names that contain it as a prefix
    val prefixLst = containsPrefix(name, lst)

    // if the name is not in the given list and has no prefixes, do not say anything. simply leave. do not say anything. pass go, do not collect $200.
    if(!prefixLst.isNotEmpty()){
        return
    }

    // print out what the user searched for
    println(name)
    // if there are names that have the prefix, then print out the information for those names
    prefixLst.forEach(){
        returnNameInfo(lst, gender, it)
    }
}

/**
    checks to see if the name is in the list. 
    @param: name (the name you're checking for), lst (list your checking in), count (to itterate)
    @return: count iff true (i.e., the name occurs in the list, so return the index for ease)
             else, -1 (an arbitray number that count probably can't be)
*/ 
fun containsName(name: String, lst: MutableList<NameHolder>, count: Int = 0): Int{
    // if you reach the end, it's not in the list and return false (-1)
    if(count > lst.lastIndex){
        return -1
    }

    // if the name is equal to the name at the index, return the index
    if(name.equals(lst[count].name, ignoreCase = true)){
        return count
    }

    return containsName(name, lst, count + 1)
}

/**
    Returns a list of indexes of where there are names with the prefix
    @param: name (name/prefix searching for), lst (list searching from)
    @return: the list of indexes that contain the name/prefix. else, return an empty list
 */ 
fun containsPrefix(name: String, lst: MutableList<NameHolder>): MutableList<Int> {
    // the list of indexes that you will return
    val retLst: MutableList<Int> = mutableListOf()

    // does the get list of indexes 
    fun checkPrefixGetIndx(name: String, lst: MutableList<NameHolder>, count: Int = 0){
        if(count > lst.lastIndex){
            return
        }
        if(lst[count].name.length >= name.length){
            val prefixName = (lst[count].name).substring(0, name.length)
            // println(prefixName)
            if(prefixName.contains(name, ignoreCase = true)){
                // println("check")
                retLst.add(count)
            }
        }
        return checkPrefixGetIndx(name, lst, count + 1)
    }
    checkPrefixGetIndx(name, lst)
    // println(retLst)
    return retLst
}

/**
    Adds an item in a list so that it preserves an alphabetical order
    @param nameChunk (the name that wants to be checked, and the item to be added), lst (the list to be added to)
*/
fun addAlphabetically(nameChunk: NameHolder, lst: MutableList<NameHolder>, count: Int = 0){
    // if you reach the end, add the name at the end
    if(count > lst.lastIndex){
        lst.add(nameChunk)
        return
    }
    // println(name.compareTo(lst[count]))
    // println(count)

    // if the name is alphabetically before a name at a given index, put it before (i.e. at, insert) that index
    if(nameChunk.name.compareTo(lst[count].name) < 0){
        lst.add(count, nameChunk)
        return
    }
    // otherwise, KEEP GOING
    return addAlphabetically(nameChunk, lst, count + 1)
}

/**
    A function that returns information about a given name (aka a more fancy toString but function)
    @param: lst (that you're getting information about the name from), gender (gender of the name), indx (the index of the name)
 */
fun returnNameInfo(lst: MutableList<NameHolder>, gender: String, indx: Int){
    // get the denomenator for the calc of how many kids are in the lists
    val totalKids = lst.map{it.totalAmount}.fold(0) {total, next -> total + next }

    // calculate the percentage
    val percent = ((lst[indx].totalAmount.toDouble()) / (totalKids.toDouble())) * 100.0

    val rank = indx + 1
    
    // print out the findings
    // println(lst[indx].name) // to match the formating on the site
    println(String.format("%s: %s \t %s: %s \t %s: %d", "Name", lst[indx].name, "Gender",gender, "# of Years",lst[indx].years))
    println(String.format("%s: %d", "Alphabetical Position", rank))
    println(String.format("%s: %d", "Occurences", lst[indx].totalAmount)) 
    println(String.format("%s %s %s: %f","Percent of All", gender, "Names", percent))
    println() // because FORMATING
}

fun main(args:Array<String>){
    // "do it so it's mutable only w/in a function" geoff i'm not that clever idk how to do that
    val BoyNames: MutableList<NameHolder> = mutableListOf()
    val GirlNames: MutableList<NameHolder> = mutableListOf()
    val NameSearchFor: MutableList<String> = mutableListOf()

    // get the user inputs
    args.forEach(){
        NameSearchFor.add(it)
    }
    
    // function to read the url
    val readURL = { url:String -> 
        val r = mutableListOf<String>();
        try {
            val oracle = URL(url);
            val br = BufferedReader(InputStreamReader(oracle.openStream()));
            loop@ while (true) {
                val line = br.readLine();
                if (line==null) {
                    break@loop
                }
                r.add(line);
            }
        } catch (ee: Exception) {
            println("Problem ${ee}");
        }
        r
    } 
    
    // would it be easier to just manually write it out? probably. am I doing that? no. please help.
    val filesLst = listOf("https://cs.brynmawr.edu/cs245/Data/names2000.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2001.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2002.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2003.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2004.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2005.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2006.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2007.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2008.csv",
        "https://cs.brynmawr.edu/cs245/Data/names2009.csv"
    )   

    // list to hold the read files
    val readFiles: MutableList<MutableList<String>> = mutableListOf()
    
    // read all the files
    filesLst.forEach(){
        readFiles.add(readURL(it))
    }

    // put all the name info into the two lists
    readFiles.forEach(){
        it.map{nameStorer(it, BoyNames, GirlNames)}
    }

    // run the search for all the names
    NameSearchFor.forEach(){
        nameSearch(it, BoyNames, "Boy")
        nameSearch(it, GirlNames, "Girl")
    }

    /*
        // test to see how it works
        val lst2000Names = readURL("https://cs.brynmawr.edu/cs245/Data/names2000.csv")
        val lst2001Names = readURL("https://cs.brynmawr.edu/cs245/Data/names2001.csv")

        // println(lst2000Names[0]) // is a string   
        lst2000Names.map{nameStorer(it, BoyNames, GirlNames)} // -> foreach or itt over for them all and then at the end: metrics
        lst2001Names.map{nameStorer(it, BoyNames, GirlNames)}

        // testiiiing 
        nameSearch("aaron", BoyNames, "Boy")
        nameSearch("dAN", BoyNames, "Boy")
        nameSearch("Marlen", GirlNames, "Girl")
        nameSearch("syd", GirlNames, "Girl")
    */
}